### PR TITLE
Code reloading with entr

### PR DESCRIPTION
**Pull request by aderyabin**
Functionality fully described [here](https://github.com/hanami/hanami/issues/249#issuecomment-181797020):

1. Remove `shotgun` as hard dependency from `hanami.gemspec`
2. Generate new projects with this gem in `Gemfile` (see below)
3. For the server: detect if `Shotgun` is present and activate code reloading via `Shotgun` (which is the actual implementation)
4. Otherwise try to fallback to `entr`
5. When none of them isn't available, it should not try to activate code reloading.

// cc @jodosha 

**NOTE**: There are no link to Hanami Guides.

### CODE REVIEW

File: lib/hanami/commands/server.rb [hunk 1]
```diff
@@ -24,12 +27,15 @@ class Server < ::Rack::Server
       # @since 0.1.0
       # @see Hanami::Environment#initialize
       def initialize(options)
+        @entr_options =  options.merge(code_reloading: false).inject([]) {|res, (k, v)| res << "--#{k}=#{v}" ; res}.join(" ")
```
Comment (jodosha) [inline lib/hanami/commands/server.rb]: @aderyabin Does this means we're open to accept any kind of CLI args and pass down to `entr`?

File: lib/hanami/commands/server.rb [hunk 2]
```diff
@@ -24,12 +27,15 @@ class Server < ::Rack::Server
...
         @_env    = Hanami::Environment.new(options)
         @options = _extract_options(@_env)
-
         if code_reloading?
+          begin
```
Comment (jodosha) [inline lib/hanami/commands/server.rb]: @aderyabin Can we please declutter `#initialize`, by extracting a method like `prepare_code_reloading!`? Thank you! :+1: This should include `@entr_options` line too. :smile:

File: lib/hanami/commands/server.rb (hunk 1 above)
Comment (aderyabin) [inline lib/hanami/commands/server.rb]: yes, all passed CLI args will be forwarded to `entr` but with disabled code reloading

File: lib/hanami/commands/server.rb [hunk 3]
```diff
@@ -17,6 +17,9 @@ module Commands
     # @since 0.1.0
     # @api private
     class Server < ::Rack::Server
+      ENTR_EXECUTE_COMMAND = 'find . | entr -r bundle exec hanami server %s'.freeze
+      ENTR_CHECK_COMMAND = 'which entr'.freeze
```
Comment (jodosha) [inline lib/hanami/commands/server.rb]: @aderyabin Using `which` makes this code non portable on Windows. Please check if this snippet can work for us: http://stackoverflow.com/a/5471032/498386

File: lib/hanami/commands/server.rb (hunk 1 above)
Comment (jodosha) [inline lib/hanami/commands/server.rb]: @aderyabin What's the need? I'm not criticizing the choice, but trying to understand the reasons behind :smile:

**Comment by jodosha** [comment]
@aderyabin This is a great implementation. Before to merge I want to try it for a development session. :+1:

**Comment by jodosha** [comment]
@aderyabin I've finally merged this. Thank you :+1:

**Comment by rbishop** [comment]
I'm just here to clap my hands in excitement to see a project using finally `entr` instead re-inventing file system polling and bringing in all sorts of crazy dependencies. :clap: :clap:

**Pull request merged**